students/ChrisH/lesson05/mailroom.py

```diff
@@ -0,0 +1,162 @@
+from collections import namedtuple
+
+# Giving namedtuples a shot
+Donor = namedtuple('Donor', 'first,last')
+
+
+# Global data structure
+donor_data = \
+    {Donor('Al', 'Donor1'): [10.00, 20.00, 30.00, 40.00, 50.00],
+     Donor('Bert', 'Donor2'): [10.00],
+     Donor('Connie', 'Donor3'): [10.00, 10.00, 10.01],
+     Donor('Dennis', 'Donor4'): [10.00, 20.00, 20.00],
+     Donor('Egbert', 'Donor5'): [10.39, 20.21, 10.59, 4000.40],
+     }
+
+
+def get_donor_fullname(donor):
+    """
+    Given a donor namedtuple, assembles and returns the donor's full name.
+    :param donor: namedtuple type: Donor
+    :return: string with donor's full name
+    """
+    return (donor.first + ' ' + donor.last).strip()   # Strip, in case first name is blank
+
+
+def menu(menu_data):
+    """
+    Prints the main user menu & retrieves user selection.
+    :param: a menu dictionary, key as numbered list.
+            Dict key 0 should be a null function (or text for bad choice)
+            Dict value is a list with the following entries:
+                [0] = Text for user prompt
+                [1] = Function that should be run for that choice
+    :return: the function corresponding to the user's selection
+    """
+    print("\nPlease choose one of the following options:")
+    for n in range(1, len(menu_data)):                     # Start at item 1, item 0 should be null function
+        print(f"{n}) {menu_data[n][0]}")                   # Prints the menu user text
+    menu_entry = menu_data.get(int(input("> ")), menu_data[0])   # if bad option, returns item 0
+    return menu_entry[1]                                         # Returns the function from that menu choice's list
+
+
+def generate_letter(donor):
+    """
+    Generates a Thank You letter to send to a donor. Uses the last value in their donations list to
+    mention their last donation amount.
+    :param donor: a donor dictionary entry
+    :return: string containing the text of the Thank You letter.
+    """
+    format_string = """
+Dear {first_name} {last_name},
+
+   Thank you for your donation of ${last_donation:.2f}.
+
+            Warmest Regards,
+                Local Charity
+"""
+    amount = donor_data[donor][-1]
+    return format_string.format(last_donation=float(amount), first_name=donor.first, last_name=donor.last)
+
+#    return "{first_name} {last_name} {donations[2]}".format(**(donor_data[donor]))  # This requires static index!!
+
+
+def send_letters_all():
+    """
+        Runs through donor data structure, generates a thank you letter for each and saves it to
+        the current working directory with in a date+donorname.txt file
+        :return: None
+        """
+    for donor in donor_data:
+        print(f'Generating letter for {get_donor_fullname(donor)}')
+        now = time.localtime()
+        f_name = f"{now.tm_year}{now.tm_mon:02}{now.tm_mday:02}_"
+        f_name += get_donor_fullname(donor).replace(" ", "_") + ".txt"
+        file_out = open(f_name, 'w')
```

Comment:
Can you use a `with` statement here instead?

Reply:
Oh, yeah! Forgot that trick for a moment. Fixed.